### PR TITLE
sql: fix wrong region allocation

### LINKED ISSUE #8763: ASAN: heap-buffer-overflow in generate_column_metadata
```console
Tarantool 3.0.0-entrypoint-256-g088b32f38
type 'help' for interactive help
```

In scope of #7327 I use `malloc` implementation of small library (details in future PR for the issue). `sql/bind.test.lua` starts to fail:

```asan
=================================================================
==2680267==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x606000050e1c at pc 0x557acc58ac09 bp 0x7f9f54eee810 sp 0x7f9f54eee808
WRITE of size 4 at 0x606000050e1c thread T0
    #0 0x557acc58ac08 in generate_column_metadata /home/shiny/dev/tarantool/src/box/sql/select.c:1956:25
    #1 0x557acc559319 in sqlSelect /home/shiny/dev/tarantool/src/box/sql/select.c:6511:3
    #2 0x557acc2dd5f6 in yy_reduce /home/shiny/dev/tarantool/src/box/sql/parse.y:421:3
    #3 0x557acc2d10cc in sqlParser /home/shiny/dev/tarantool/build-asan-debug/src/box/sql/parse.c:4424:7
    #4 0x557acc5dfa4a in sqlRunParser /home/shiny/dev/tarantool/src/box/sql/tokenize.c:577:4
    #5 0x557acc4f25d8 in sql_stmt_compile /home/shiny/dev/tarantool/src/box/sql/prepare.c:79:4
    #6 0x557acc2707f4 in sql_prepare_and_execute /home/shiny/dev/tarantool/src/box/execute.c:248:6
    #7 0x557acc888b82 in lbox_execute /home/shiny/dev/tarantool/src/box/lua/execute.c:495:7
    #8 0x557accc62426 in lj_BC_FUNCC /home/shiny/dev/tarantool/build-asan-debug/third_party/luajit/src/lj_vm.S:811
    #9 0x557accc8c398 in lua_pcall /home/shiny/dev/tarantool/third_party/luajit/src/lj_api.c:1163:12
    #10 0x557acc91e0ee in luaT_call /home/shiny/dev/tarantool/src/lua/utils.c:618:6
    #11 0x557acc8f43fc in lua_fiber_run_f /home/shiny/dev/tarantool/src/lua/fiber.c:416:11
    #12 0x557acb78abd3 in fiber_cxx_invoke(int (*)(__va_list_tag*), __va_list_tag*) /home/shiny/dev/tarantool/src/lib/core/fiber.h:1234:10
    #13 0x557acc9f8b42 in fiber_loop /home/shiny/dev/tarantool/src/lib/core/fiber.c:1002:18
    #14 0x557acd953da8 in coro_init /home/shiny/dev/tarantool/third_party/coro/coro.c:108:3

0x606000050e1f is located 0 bytes to the right of 63-byte region [0x606000050de0,0x606000050e1f)
allocated by thread T0 here:
    #0 0x557acb739f59 in malloc (/home/shiny/dev/tarantool/build-asan-debug/src/tarantool+0x1269f59) (BuildId: a15902804215b5a5ba08ce04eeadedb86c8cb49d)
    #1 0x557acd89982f in slab_get /home/shiny/dev/tarantool/src/lib/small/include/small/slab_cache_malloc.h:109:37
    #2 0x557acd89923e in region_reserve_slow /home/shiny/dev/tarantool/src/lib/small/small/region.c:43:26
    #3 0x557acc58f928 in region_reserve /home/shiny/dev/tarantool/src/lib/small/include/small/region.h:206:9
    #4 0x557acc58f098 in region_aligned_reserve /home/shiny/dev/tarantool/src/lib/small/include/small/region.h:232:14
    #5 0x557acc53f887 in region_aligned_alloc /home/shiny/dev/tarantool/src/lib/small/include/small/region.h:240:14
    #6 0x557acc58a07b in generate_column_metadata /home/shiny/dev/tarantool/src/box/sql/select.c:1939:3
    #7 0x557acc559319 in sqlSelect /home/shiny/dev/tarantool/src/box/sql/select.c:6511:3
    #8 0x557acc2dd5f6 in yy_reduce /home/shiny/dev/tarantool/src/box/sql/parse.y:421:3
    #9 0x557acc2d10cc in sqlParser /home/shiny/dev/tarantool/build-asan-debug/src/box/sql/parse.c:4424:7
    #10 0x557acc5dfa4a in sqlRunParser /home/shiny/dev/tarantool/src/box/sql/tokenize.c:577:4
    #11 0x557acc4f25d8 in sql_stmt_compile /home/shiny/dev/tarantool/src/box/sql/prepare.c:79:4
    #12 0x557acc2707f4 in sql_prepare_and_execute /home/shiny/dev/tarantool/src/box/execute.c:248:6
    #13 0x557acc888b82 in lbox_execute /home/shiny/dev/tarantool/src/box/lua/execute.c:495:7
    #14 0x557accc62426 in lj_BC_FUNCC /home/shiny/dev/tarantool/build-asan-debug/third_party/luajit/src/lj_vm.S:811

SUMMARY: AddressSanitizer: heap-buffer-overflow /home/shiny/dev/tarantool/src/box/sql/select.c:1956:25 in generate_column_metadata
Shadow bytes around the buggy address:
  0x0c0c80002170: fa fa fa fa fd fd fd fd fd fd fd fd fa fa fa fa
  0x0c0c80002180: fd fd fd fd fd fd fd fd fa fa fa fa fd fd fd fd
  0x0c0c80002190: fd fd fd fd fa fa fa fa fd fd fd fd fd fd fd fd
  0x0c0c800021a0: fa fa fa fa fd fd fd fd fd fd fd fd fa fa fa fa
  0x0c0c800021b0: fd fd fd fd fd fd fd fd fa fa fa fa 00 00 00 00
=>0x0c0c800021c0: 00 00 00[07]fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0c800021d0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0c800021e0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0c800021f0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0c80002200: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0c80002210: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==2680267==ABORTING
```

### PR DESCRIPTION
**Pull request by ImeevMA**
This patch fixes an issue in generate_column_metadata(). Prior to this patch, the number of variable-only expressions was counted incorrectly when temporary memory was allocated on region to store their positions. However, although this allocation was incorrect, this did not lead to any problems due to the specifics of the region allocations.

This patch fixes this by removing the temporary memory allocation.

Closes #8763

NO_DOC=no user-visible changes
NO_TEST=no user-visible changes
NO_CHANGELOG=no user-visible changes

### CODE REVIEW

File: src/box/sql/select.c [hunk 1]
```diff
@@ -1934,14 +1934,6 @@ generate_column_metadata(struct Parse *pParse, struct SrcList *pTabList,
 	if (pParse->colNamesSet)
```
Comment (Gerold103) [inline src/box/sql/select.c:1934]: Thanks for the patch!

Could you please elaborate what was wrong here exactly. I can't see any problems in the old code.

Comment (ImeevMA) [reply src/box/sql/select.c:1934]: In old code memory was allocated according to number of variables, which can be different from number of variables uses, for example:
```
box.execute([[SELECT $1, $1, $1;]], {123})
```
Here we have only one variable, but we need three sets of metadata.

**Review by Gerold103 — APPROVED** [review]
Thanks for the explanation!

File: src/box/sql/select.c (hunk 1 above)
Comment (Gerold103) [reply src/box/sql/select.c:1934]: Oh, I see, indeed.

**Review by igormunkin — APPROVED** [review]
@ImeevMA, thanks for the patch! LGTM.

**Pull request merged**